### ISSUE TITLE
WFL limited usage, way to access WML variables and compare different kinds of objects

### ISSUE THREAD

**Issue by hladkyjakub** [wesnoth/wesnoth#9039]
### Describe the desired feature

Hi,
I'm a WML/Lua modder, currently working on some abilities, in the process I've noticed WFL cannot work with global WML variables(like turn_number) (which is bad even for standard unit filtering, because you can't, for example, filter if variable stored in the unit is smaller or equal than turn number).
Also, WFL cannot compare the current unit and current weapon at once in [ability] (tho it can access both of them).
...
So I think it would be nice if the working space of WFL could be extended so it's more usable than WML with many formulas and variable substitutions.

I am unsure about the Feature Request format, so if I missed something or messed up the formatting, please inform me.

**Comment by CelticMinstrel** [comment]
Can you be more specific? I believe several of the statements in your post are inaccurate. For example, it's definitely possible to check the current turn number in _some_ WFL formulas. Though I haven't checked the source code, I was also under the impression that all weapon special formulas exposed variables for both units _and_ both weapons, though it wouldn't surprise me at all if this isn't properly documented.

In short, I need to know _where_ you are using a formula that cannot access whatever you want to access. In other words, which formula is unable to access the turn number? The variables a formula can access depend entirely upon where that formula is used.

There doubtless are some holes, even if they may not be the exact ones you mentioned, so certainly there's a lot of room for improvement with regard to WFL's utility.

**Comment by hladkyjakub** [comment]
> Can you be more specific? I believe several of the statements in your post are inaccurate. For example, it's definitely possible to check the current turn number in _some_ WFL formulas. Though I haven't checked the source code, I was also under the impression that all weapon special formulas exposed variables for both units _and_ both weapons, though it wouldn't surprise me at all if this isn't properly documented.

True, but it could be mentioned more than just in the link https://wiki.wesnoth.org/EventWML#filter_formula 
(Only precise read-through now revealed this information)... great to know actually.

... Sorry for myself, but how to access them? I kinda cannot do that (In [resistance], due to syntax probably, but nowhere is mentioned the right syntax)

> In short, I need to know _where_ you are using a formula that cannot access whatever you want to access. In other words, which formula is unable to access the turn number? The variables a formula can access depend entirely upon where that formula is used.

Anywhere they cannot access WML variables.

> There doubtless are some holes, even if they may not be the exact ones you mentioned, so certainly there's a lot of room for improvement with regard to WFL's utility.

: )  Actually, I started to doubt if game state isn't accessible only in [event]?

**Comment by CelticMinstrel** [comment]
I'm not looking for broad generalizations but rather for specific, precise instances.

I took the time since my last post to verify that indeed the ability formulas have access to both the unit and the weapon… but just from your imprecise words, I'm not sure you're even talking about what I'm calling the ability formulas. The formulas I'm referring to as "ability formulas" are any formula used in a key _directly_ within an ability tag… so, it excludes formulas in any of the filter tags nested in the ability.

To identify a particular formula that you claim has a shortcoming, I need to know:

* The general context of where the formula was used (eg: abilities, events, etc)
* The exact name of the WML key that contains the formula
* The exact name of the WML tag that contains the above key
* Possibly, the name of the tag that contains the above tag could also help, and the name of the tag that contains that tag, etc. Most of the time it's probably not necessary, but…

So, "anywhere they cannot access WML variables" is not a reasonable identification.

**Comment by hladkyjakub** [comment]
1. You can't access my_variable from any formula.
2. I was talking about formulas directly in [resistance]/[damage]... tag.
Q1. You can put formula directly into [abilities]?
Q2. I still coudn't find how the syntax for gamestate events works (if they can be even used there)... do you know this?

**Comment by CelticMinstrel** [comment]
> You can't access my_variable from any formula.

I don't know what you mean by my_variable.

> I was talking about formulas directly in [abilities][resistance]/[specials][damage] etc. tag.

You're saying formulas in, eg, `[specials][damage]` cannot access both the unit and the weapon? I don't think that's true though? From the source code, it looks like the weapon should be `self` and the units should be `attacker` and `defender`. There may be more than that too, as I didn't look extremely closely at the source code.

> You can put formula directly into [abilities]?

If you mean attributes directly in the `[abilities]` tag, no. That's not supported. Only tags may go in `[abilities]`. But you can put formulas in `[abilities][leadership]damge=` for example.

> I still coudn't find how the syntax for gamestate events works (if they can be even used there)... do you know this?

I don't understand the question. Is it even about WFL?

> One more uselfull thing: other tags (non-numerical) in abilities/specials could also support formulas (like apply_to=)

That doesn't seem all that useful to me, but… I guess there could be niche use-cases.

**Comment by soliton-** [comment]
> From the source code, it looks like the weapon should be `self` and the units should be `attacker` and `defender`.

Would be good to update https://wiki.wesnoth.org/AbilitiesWML#Available_formula_variables_in_Abilities_and_Weapon_Specials with that info.
